src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala

```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
+
+  /** @return lines formatted as: <customer_name> has to pay <total_price> */
   def report2(orders: List[Order], products: List[Product], customers: List[Customer]): List[String] = {
-    orders.groupBy(_.customerId)
-      .map { case (customerId, ordersOfCustomer) =>
-        val totals = ordersOfCustomer
-          .map(order => {
-            products.find(product => product.id == order.productId)
-              .map(_.price * order.amount)
-          })
-
-        val reportLine = for {
-          customer <- customers.find(_.id == customerId)
-          t <- if (totals.exists(_.isEmpty)) Option.empty else Option(totals.map(_.get).sum)
-        } yield s"${customer.firstName} ${customer.lastName} has to pay $t"
-
-        reportLine.getOrElse(s"CORRUPT ORDERS for customer $customerId")
-      }
-      .toList
-  }
+    for {
+      person <- LineTotal(orders, products, customers).groupBy(rec => (rec.firstName, rec.lastName))
+      (firstName, lastName) = person._1
+      total = person._2.map(_.total).sum
+    } yield s"$firstName $lastName has to pay $total"
+  }.toList
 
   def generateReports(customerInput: BufferedInputStream,
                       productInput: BufferedInputStream,
                       orderInput: BufferedInputStream): (List[String], List[String]) = {
     val orders = readOrders(orderInput)
     val products = readProducts(productInput)
     val customers = readCustomers(customerInput)
-
-    val r1 = report1(orders, products, customers)
-    val r2 = report2(orders, products, customers)
-
-    (r1, r2)
+    val wants = report1(orders, products, customers)
+    val hasToPay = report2(orders, products, customers)
+    (wants, hasToPay)
   }
 
-  def writeReport(report: List[String], output: BufferedWriter): Unit = {
+  def writeReport(report: List[String], output: BufferedWriter): Unit =
     output.write(report.mkString("\n"))
-  }
 
   val reports: ManagedResource[(List[String], List[String])] = for {
-    cIn <- Using.fileInputStream(customerFile)
-    pIn <- Using.fileInputStream(productFile)
-    oIn <- Using.fileInputStream(orderFile)
-  } yield generateReports(cIn, pIn, oIn)
+    orders <- Using.fileInputStream(orderFile) //.acquireAndGet(FileIO.readOrders)
+    customers <- Using.fileInputStream(customerFile) //.acquireAndGet(FileIO.readCustomers)
+    products <- Using.fileInputStream(productFile) //.acquireAndGet(FileIO.readProducts)
+  } yield generateReports(customers, products, orders)
 
+  import resource._
   val reportWriting: ManagedResource[Unit] = for {
-    rs <- reports
-    (report1, report2) = rs
-    r1 <- Using.fileWriter()(report1File)
-    r2 <- Using.fileWriter()(report2File)
-  } yield {
-    writeReport(report1, r1)
-    writeReport(report2, r2)
-  }
+    reps <- reports
+    writer1 <- managed(new FileWriter(new File("target/report1.txt")))
+    writer2 <- managed(new FileWriter(new File("target/report2.txt")))
+    bufferedWriter1 <- managed(new BufferedWriter(writer1))
```

Comment:
I see why you would do both the `FileWriter` and the `BufferedWriter`, but isn't this just what `fileWriter` is already doing?